source/fe/mapping_fe.cc

```diff
@@ -0,0 +1,1385 @@
+
+      return;
+    }
+
+    default:
+      Assert(false, ExcNotImplemented());
+    }
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+template < int rank >
+void MappingFE<dim,spacedim,DH,VECTOR>::transform_differential_forms(
+  const VectorSlice<const std::vector<DerivativeForm<rank, dim,spacedim> > >    input,
+  VectorSlice<std::vector<DerivativeForm<rank, spacedim,spacedim> > > output,
+  const typename Mapping<dim,spacedim>::InternalDataBase &mapping_data,
+  const MappingType mapping_type) const
+{
+
+  AssertDimension (input.size(), output.size());
+  Assert (dynamic_cast<const InternalData *>(&mapping_data) != 0,
+          ExcInternalError());
+  const InternalData &data = static_cast<const InternalData &>(mapping_data);
+
+  switch (mapping_type)
+    {
+    case mapping_covariant:
+    {
+      Assert (data.update_flags & update_contravariant_transformation,
+              typename FEValuesBase<dim>::ExcAccessToUninitializedField("update_contravariant_transformation"));
+
+      for (unsigned int i=0; i<output.size(); ++i)
+        output[i] = apply_transformation(data.covariant[i], input[i]);
+
+      return;
+    }
+    default:
+      Assert(false, ExcNotImplemented());
+    }
+
+}
+
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+Point<spacedim>
+MappingFE<dim,spacedim,DH,VECTOR>::
+transform_unit_to_real_cell (const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+                             const Point<dim>                                 &p) const
+{
+//  Use the get_data function to create an InternalData with data vectors of
+//  the right size and transformation shape values already computed at point
+//  p. Make sure only one processor at a time performs this operations.
+
+  Threads::Mutex::ScopedLock lock(mutex);
+  update_internal_dofs(cell);
+  const Quadrature<dim> point_quadrature(p);
+  std::auto_ptr<InternalData>
+  mdata (dynamic_cast<InternalData *> (
+           get_data(update_transformation_values, point_quadrature)));
+
+  return this->transform_unit_to_real_cell_internal(*mdata);
+}
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+Point<spacedim>
+MappingFE<dim,spacedim,DH,VECTOR>::
+transform_unit_to_real_cell_internal (const InternalData &data) const
+{
+  Point<spacedim> p_real;
+
+  for (unsigned int i=0; i<data.n_shape_functions; ++i)
+    {
+      unsigned int comp_i = fe->system_to_component_index(i).first;
+      if (fe_mask[comp_i])
+        p_real[fe_to_real[comp_i]] += local_dofs[i] * data.shape(0,i);
+    }
+
+  return p_real;
+}
+
+
+
+template<int dim, int spacedim, class DH, class VECTOR>
+Point<dim>
+MappingFE<dim,spacedim,DH,VECTOR>::
+transform_real_to_unit_cell (const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+                             const Point<spacedim>                            &p) const
+{
+  {
+    Threads::Mutex::ScopedLock lock(mutex);
+    update_internal_dofs(cell);
+  }
```

Comment:
here and in similar places below, this won't work: you only lock the update operation, but this operation sets certain member variables that I assume you want to use below somewhere (I assume -- I don't know). But because the lock is immediately released again, another thread may overwrite these member variables as soon as you get past the closing brace here.